keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,553 @@
+
+#### Local-Opt
+LO shares the same idea of `recommended` that rebalances endpoints from
+overflowed zones to insufficient zones. Instead of keeping more endpoints than
+expected in a `local EndpointSlice Group`, it shares those overall extra
+endpoints to all zones with an evenly divided weight to achieve better
+overload.
+
+##### Rationale
+1. Relatively high in-zone traffic
+2. Relatively balanced deviation
+
+##### Potential Issues
+1. Weights involved in the tie-breaker stage
+   - Requires API addition
+   - Introduces more volatility
+2. Higher number of EndpointSlices
+3. Lower in-zone traffic than `recommended`
+
+
+### Test Plan
+| Test Name | Test Type | Test Details | Expected Output | Comments |
+| :--- | :--- | :--- | :--- | :--- |
+| Single Approach Test | Unit Test | Small/Large number of nodes | Distribute EndpointSlices as expected | |
+| |  | Balanced/Imbalanced endpoints | Distribute EndpointSlices as expected | |
+| |  | Zero endpoints in some zones | Distribute EndpointSlices as expected | |
+| |  | Invalid annotation | Approach sets to default, distribute EndpointSlices as expected | |
+| `Prefer` Test | Unit Test | Invalid starting threshold | Set starting threshold to default and start `Prefer` approach with condition meets| This is applicable for customized starting threshold only|
+| |  | Endpoints below `starting threshold + padding` | `Prefer` approach will not be conducted | |
+| |  | Endpoints above `starting threshold + padding` | `Prefer` works as expected | |
+| Approach Transition Test | Unit Test | 1.Set annotation to `Prefer` <br>2. Increase endpoints number to above `starting threshold` but below `starting threshold + padding` | `Prefer` will not be conducted | If capable, monitor the churn of EndpointSlices during the rolling update |
+|  |  | 1.Set annotation to `Prefer` <br>2. Increase endpoints number to above `starting threshold + padding` | Distribution will be changed from `Balanced` to `Prefer` | Same as above |
+|  |  | 1.Set annotation to `Prefer` <br>2. Decrease endpoints number to below `starting threshold` but above `starting threshold - padding` | Distribution will remain `Prefer` | Same as above |
+|  |  | 1.Set annotation to `Prefer` <br>2. Decrease endpoints number to below `starting threshold - padding` | Distribution will be changed from `Prefer` to `Balanced` | Same as above |
+|  |  | 1.Set annotation to one of the three <br>2. Set endpoints to above `starting threshold + padding` <br> 3. Manually change the annotation to another approach | Distribution will be changed based on the new value | Step 2 is only needed for `Prefer` |
+| Rolling Update Test | E2E Test |1.Set annotation to `Prefer` <br>2. Conduct an aggressive rolling update policy for endpoints update <br> 3. Set endpoints to test corner cases | In the desired state, EndpointSlices are distributed as expected | If capable, monitor the churn of EndpointSlices during the rolling update |
```

Comment:
I'm wondering if the integration framework will be better suited for some of these tests,  since we are stressing only the endpointslice controller and we can mock the pods, it will allow us to create larger topologies and to instrument the test so we can calculate the churn rate, .... and fail the test on those parameters

Reply:
This makes sense. Thanks for the feedback, updated!